test/spec.ml

```diff
@@ -0,0 +1,224 @@
+  ++ hash
+
+type inode_value = tree_entry list
+
+let inode_value v =
+  assert (List.length v <= 32);
+  "\000"
+  ++ leb128_int (List.length v)
+  ++ list inode_entry (List.sort (fun e e' -> String.compare e.name e'.name) v)
+
+type inode_pointer = { index : int; hash : hash }
+
+let inode_pointer { index; hash } =
+  assert (index <= 32);
+  leb128_int index ++ hash
+
+type inode_tree = {
+  depth : int;
+  entries_length : int;
+  pointers : inode_pointer list;
+}
+
+let inode_tree { depth; entries_length; pointers } =
+  assert (List.length pointers <= 32);
+  "\001"
+  ++ leb128_int depth
+  ++ leb128_int entries_length
+  ++ leb128_int (List.length pointers)
+  ++ list inode_pointer
+       (List.sort (fun p p' -> Int.compare p.index p'.index) pointers)
+
+(** {4 Internal tree construction} *)
+
+let ( ** ) = Int32.mul
+let ( ^= ) a b = a := Int32.logxor !a b
+let ( *= ) a b = a := !a ** b
+let ( << ) = Int32.shift_left
+let ( >> ) = Int32.shift_right_logical
+let ( ||| ) = Int32.logor
+let ( &&& ) = Int32.logand
+
+let mix h w =
+  w *= 0xcc9e2d51l;
+  w := !w << 15 ||| (!w >> 17);
+  w *= 0x1b873593l;
+  h ^= !w;
+  h := !h << 13 ||| (!h >> 19);
+  h := Int32.add (!h ** 5l) 0xe6546b64l
+
+let last s off =
+  Bytes.init 4 (fun i ->
+      let j = off + i in
+      if j >= String.length s then '\000' else s.[j])
+
+let ocaml_hash seed s =
+  let h = ref (Int32.of_int seed) in
+  let w = ref 0l in
+  let rec loop i =
+    if i + 4 > String.length s then i
+    else (
+      w := Bytes.get_int32_le (Bytes.unsafe_of_string s) i;
+      mix h w;
+      loop (i + 4))
+  in
+  let i = loop 0 in
+  if i <> String.length s then (
+    w := Bytes.get_int32_le (last s i) 0;
+    mix h w);
+  h ^= Int32.of_int (String.length s);
+  h ^= (!h >> 16);
+  h *= 0x85ebca6bl;
+  h ^= (!h >> 13);
+  h *= 0xc2b2ae35l;
+  h ^= (!h >> 16);
+  Int32.to_int (!h &&& 0x3FFFFFFFl)
+
+let index d n = ocaml_hash d n mod 32
+
+type inode = Empty | Value of inode_value | Tree of inode_tree
+
+(** This implements the computation of $X_{d,j}$. *)
+let filter x d j = List.filter (fun { name; _ } -> index d name = j) x
+
+let rec partition d x =
+  match List.length x with
+  | 0 -> Empty
+  | n when n <= 32 -> Value x
+  | n ->
+      let tp =
+        List.init 32 (fun j ->
+            let tj = partition (succ d) (filter x d j) in
+            let tj_prehash =
+              match tj with
+              | Empty -> inode_value []
+              | Value v -> inode_value v
+              | Tree t -> inode_tree t
+            in
+            (tj, { index = j; hash = blake2b tj_prehash }))
+      in
+      let p =
+        List.filter_map
+          (fun (ti, pi) -> match ti with Empty -> None | _ -> Some pi)
+          tp
+      in
+      Tree { depth = d; entries_length = n; pointers = p }
+
+let inode x =
+  match partition 0 x with
+  | Value v -> inode_value v
+  | Tree t -> inode_tree t
+  | Empty -> (* This should not happen *) inode_value []
```

Comment:
```suggestion
  | Empty -> (* This should not happen *)
```

If so, then why not `assert false`?

Reply:
The spec gives a semantics to this, so I just implemented it as close to it as possible.
I don't mind changing it to `assert false` if it makes it clearer